src/hydro_system.hpp

```diff
@@ -945,6 +957,19 @@ void HydroSystem<problem_t>::ComputeFluxes(amrex::MultiFab &x1Flux_mf, amrex::Mu
 		const double v_norm = (F[density_index] >= 0.) ? (F[density_index] / rho_R) : (F[density_index] / rho_L);
 		x1FaceVel(i, j, k) = v_norm;
 
+		// use the same logic as above to scale and conserve specie fluxes
+		if (F[density_index] >= 0.) {
+			for (int n = 0; n < nmscalars_; ++n) {
+				const int nstart = nvar_ - nscalars_;
+				F[nstart + n] = F[density_index] * U_L[nstart + n] / fluxSum_U_L;
+			}
+		} else {
+			for (int n = 0; n < nmscalars_; ++n) {
+				const int nstart = nvar_ - nscalars_;
+				F[nstart + n] = F[density_index] * U_R[nstart + n] / fluxSum_U_R;
+			}
+		}
+
```

Comment:
This looks right. Can you add a test problem from the CMA paper to test it?

Reply:
I am confused whether the fluxes of mass scalars should be scaled with the density_flux or not, in Quokka. Before my changes to this file, `F[MassScalar]` would just be `F[MassScalar] += viscosity * (U_L[MassScalar] - U_R[MassScalar])`, and there was no scaling with the density flux, right? 

I am trying to remember if in FLASH, before I implemented the flux conservation step, were the fluxes of mass scalars already being scaled by the density flux or not. I think they were, which is why the conservation step also included the density fluxes. But in Quokka, that doesn't seem to be the case? So, should the line `F[nstart + n] = F[density_index] * U_L[nstart + n] / fluxSum_U_L` in fact just be `F[nstart + n] = U_L[nstart + n] / fluxSum_U_L`?

Reply:
I think it is correct as you've written it. It is actually scaled by the density flux in the current version of the code.

The term you've highlighted is the artificial viscosity term, not the actual advection term. The latter is computed inside the `HLLC` function in HLLC.hpp, where there is a scaling by the density flux.

Reply:
Ah okay that's clear. But another question: if scalar fluxes are scaled by density flux in `HLLC.hpp`, and now (for mass scalars) we also scale them in `hydro_system.hpp` following the CMA implementation, are we actually scaling the mass scalar fluxes by density**2 (once in hydro_system and then again in HLLC)?

Reply:
No, you are overwriting the fluxes they were assigned earlier.